### DISCUSSION TITLE
Typing theme component properties

### DISCUSSION THREAD

**Discussion by samturrell** [chakra-ui/chakra-ui#8483] [Q&A] [answered]
Hi there!

I'm looking into how to better type the component configs when modifying the theme.

With this example i get the correct typing for the component, but it requires that you provide all of the current properties, and doesn't allow additional properties:

```
import { theme } from '@chakra-ui/react';

export const Input: (typeof theme)['components']['Input'] = {
    baseStyle: {
        field: {
            _
        },
    },
};
```

![image](https://github.com/chakra-ui/chakra-ui/assets/5918348/9b47bc65-6753-49ee-80a6-03dface27cf5)

![image](https://github.com/chakra-ui/chakra-ui/assets/5918348/7d866ab4-593c-47f4-8511-194c1afa43d6)

With the following i get full typehinting, but not specific to a single component

```
import { type ComponentStyleConfig } from '@chakra-ui/react';

export const Input: ComponentStyleConfig = {
    baseStyle: {
        field: {},
    },
};
```

Is there a way to get the best of both worlds here? I.e. provide the shape of the component config, but also allow the additional styling props to be defined?

**Comment by segunadebayo** [comment, marked as answer]
`ComponentStyleConfig` and `theme.components.Input` are v2. v2 is maintenance-only.

In v3, use `defineRecipe` / `defineSlotRecipe` and run `npx @chakra-ui/cli typegen`. Extra variants type-check after that.
